### ISSUE TITLE
[Async TP] Activations not cleared after backward when reduce_scatter_tensor saved for backward by per op SAC

### ISSUE THREAD

**Issue by danielvegamyhre** [pytorch/pytorch#149876]
## Context

This is a follow up to the discussion here: https://github.com/pytorch/torchtitan/pull/965#issuecomment-2744476861

## Summary 

After using the partitioner change which removes saved collective results which are not actually used for backward (https://github.com/pytorch/pytorch/pull/149652) and then adding back `reduce_scatter_tensor` to the [list of ops to save](https://github.com/pytorch/torchtitan/blob/f3943ddf7a9d5584a2bd3fdf9be3f1816a85bd7f/torchtitan/models/llama/parallelize_llama.py#L226) in per op SAC in torchtitan, I confirmed it works (i.e., doesn't crash) - this is awesome!
 
However, there is now a new problem to grapple with.

If we add back `reduce_scatter_tensor` to the torchtitan save list for per op SAC, the reduce scatter nodes in the graph now have 2 users instead of 1.

The async TP pattern matcher for finding subgraphs to fuse into fused_matmul_reduce_scatter nodes only matches when reduce_scatter has only 1 user (CallFunction defaults to _users=1): https://github.com/pytorch/pytorch/blob/1e159db57c611b98a531341927b2d01f39383f7a/torch/_inductor/fx_passes/micro_pipeline_tp.py#L225-L230

This means that the reduce_scatters saved for backward are not fused, because those nodes now have 2 users. Since the pattern doesn't match, no fusion occurs for those nodes.

To solve this, I tried adding some additional patterns to match single user and multi user reduce scatters: https://github.com/pytorch/pytorch/pull/149875

The fusion does indeed occur now, but there is something occuring that looks almost like a memory leak  - memory usage increases every step until OOM around step 20 (see logs: https://www.internalfb.com/phabricator/paste/view/P1765166464)

To me this indicates something like activations never getting cleared after backward, so i looked at the memory snapshot and this is indeed the case, i can see tensors allocated during the forward passes that are never freed through the rest of the training run: 

<img width="1451" alt="Image" src="https://github.com/user-attachments/assets/3b5c7131-fee4-4a7c-9635-22603cecf912" />

### Versions

Pytorch: `scatter-dim` branch (https://github.com/pytorch/pytorch/pull/149247) with @bdhirsh's PR patched in https://github.com/pytorch/pytorch/pull/149652

torchtitan: main@HEAD, with one change - saving reduce scatter tensor in per op SAC.

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @soulitzer @ezyang @albanD @gqchen @pearu @nikitaved @Varal7 @xmfan

**Comment by bdhirsh** [comment]
do you have a tlparse output from one of those runs you can share?

**Comment by danielvegamyhre** [comment]
> do you have a tlparse output from one of those runs you can share?

Here is the tlparse output: https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/.tmpZBziOw/rank_0/index.html?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=10000

**Comment by danielvegamyhre** [comment]
Seems like a lot of the problematic tensors not getting cleaned up are from [checkpoint_wrapper.forward](https://github.com/pytorch/pytorch/blob/e77ca199997c95c7fe85e606d2c226d45a866394/torch/distributed/algorithms/_checkpoint/checkpoint_wrapper.py#L171):

![Image](https://github.com/user-attachments/assets/09223586-ac61-445a-b72e-90c52e58f7d2)

**Comment by bdhirsh** [comment]
So I think there is something fishy going on in your pattern match. This snippet that fixes the leak for me:
```
diff --git a/torch/_inductor/fx_passes/micro_pipeline_tp.py b/torch/_inductor/fx_passes/micro_pipeline_tp.py
index ef42a015369..4a4e2d815b0 100644
--- a/torch/_inductor/fx_passes/micro_pipeline_tp.py
+++ b/torch/_inductor/fx_passes/micro_pipeline_tp.py
@@ -212,14 +212,14 @@ class _ReduceScatterMatch:

     def erase(self) -> None:
         for node in reversed(self.match.nodes):
-            if len(node.users) == 0:
+            if len(node.users) == 0 and node.target is not torch.ops._c10d_functional.wait_tensor.default:
                 node.graph.erase_node(node)


 def find_reduce_scatter_patterns(graph: torch.fx.Graph):
```

It's probably not quite right though. I imagine that we should only "avoid removing the wait_tensor call if its input arg is **not** part of the match.

So this is the code I see in the forward graph before inductor's pattern matching:
```
reduce_scatter_tensor: "bf16[1, 4096, 4096][16777216, 4096, 1]cuda:0" = torch.ops._c10d_functional.reduce_scatter_tensor.default(cat_1, 'sum', 2, '3');  cat_1 = None
        wait_tensor_2: "bf16[1, 4096, 4096][16777216, 4096, 1]cuda:0" = torch.ops._c10d_functional.wait_tensor.default(reduce_scatter_tensor)
...
return (..., reduce_scatter_tensor, ....)
```

And after, I get:
```
reduce_scatter_tensor: "bf16[1, 4096, 4096][16777216, 4096, 1]cuda:0" = torch.ops._c10d_functional.reduce_scatter_tensor.default(cat_1, 'sum', 2, '3');  cat_1 = None
...
return (..., reduce_scatter_tensor, ....)
```

It looks like a **different** reduce_scatter was matched and converted to an asyncTP call, but this particular node `reduce_scatter_tensor` was not actually removed. However, the corresponding `wait_tensor_2` node **was** removed

This is probably not good: in particular, since we are returning from our compiled forward function now before syncing, then we do not actually know when the input to the reduce_scatter is free to have its memory freed.

**Comment by danielvegamyhre** [comment]
>It looks like a different reduce_scatter was matched and converted to an asyncTP call, but this particular node reduce_scatter_tensor was not actually removed. However, the corresponding wait_tensor_2 node was removed

>This is probably not good: in particular, since we are returning from our compiled forward function now before syncing, then we do not actually know when the input to the reduce_scatter is free to have its memory freed.

Great find! That makes sense. 

> This snippet that fixes the leak for me: ... It's probably not quite right though. I imagine that we should only "avoid removing the wait_tensor call if its input arg is not part of the match.

I think the next step for me is to figure out why the other reduce scatter isn't getting fused properly, rather than avoid removing that wait_tensor call for the un-fused reduce scatter. When I look at the graph it seems like it should match, but it isn't getting replaced for some reason... I'll look into this - thanks!

**Comment by danielvegamyhre** [comment]
Update - I see 2 interesting things:

1) In a given input graph with 2 scaled mm reduce-scatter patterns, when using per op SAC policy that includes the `reduce_scatter_tensor` op in the save list, for some reason only 1 of the 2 reduce-scatter nodes has the final return/output node as a 2nd user (to save for backward). 

2) The reduce-scatter with 1 user is replaced properly. The reduce-scatter with 2 users is NOT replaced properly. Specifically, the fused node is inserted into the graph, but the original subgraph is not erased (see screenshot below). Stepping through with the debugger, it seems they aren't getting erased because they still have > 0 users, and the [code](https://github.com/pytorch/pytorch/blob/2a9e73783966bd6ffb75eedaafc77b38106dee4f/torch/_inductor/fx_passes/micro_pipeline_tp.py#L215) only erases nodes with 0 users. 

I think the reason is because all users of the wait_tensor node are replaced with the fused_node [here](https://github.com/pytorch/pytorch/blob/2a9e73783966bd6ffb75eedaafc77b38106dee4f/torch/_inductor/fx_passes/micro_pipeline_tp.py#L211), but since the reduce_scatter has a 2nd user (the return/output node, to save for backward), it is never erased [here](https://github.com/pytorch/pytorch/blob/2a9e73783966bd6ffb75eedaafc77b38106dee4f/torch/_inductor/fx_passes/micro_pipeline_tp.py#L215).
 
![Image](https://github.com/user-attachments/assets/01587ba5-db05-4990-b899-d6c505b922af)

**Comment by danielvegamyhre** [comment]
Update: I figured out a solution, created https://github.com/pytorch/pytorch/pull/149875

**Comment by gnadathur** [comment]
cc: @fegin
